### ISSUE TITLE
Help with training

### ISSUE THREAD

**Issue by rohansinghjain** [milesial/Pytorch-UNet#208]
Hi, I'm training on my custom dataset of 10,000 rgb images with corresponding binary masks (only one class). I have made the required changes to the preprocess function in dataset.py to ensure that the loaded masks are binary. I'm able to run train.py successfully but the results are not very encouraging.

1. On training with a small subset of 50 images (40 training and 10 validation) with a learning rate of pow(10,-6), batch size of 2, scale factor of 1, up convolution of bilinear upscaling for 50 epochs. I see that the validation coefficient stays very low throughout (~order of pow(10,-3)). Batch loss is also very high (15 to 25). I'm not able to overfit the network on a small dataset. 

2. On training with all 10,000 images with validation fraction = 20%, learning rate = pow(10,-5), epochs = 50, scale = 1 and batch size = 1,  validation dice coefficient reaches a maximum of 0.62 and keeps oscillating around 0.62. It doesn't go beyond 0.64. Batch loss fluctuates a lot between 0.2 and 0.7. The predicted results also don't look good.

Could you please suggest training hyperparameters like batch size, learning rate, epochs to be used for a dataset as big as 10,000 images to  get good results? Also, what could be the reason for not being able to overfit on a small dataset?

**Comment by milesial** [comment]
Hi, 
1. merge the overfit issue with this one https://github.com/milesial/Pytorch-UNet/issues/165
2. On the carvana dataset I used a lr of 2e-6, batch size 1 and from 1 to 5 epochs. It may be different for you, I suggest doing grid-search or random-search on the hyperparameters. Also, did you look at your training loss curve to see if your lr is too low? Last, you may try to change the lr scheduler that might reduce your lr too much.

**Comment by karlita101** [comment]
> Hi, I'm training on my custom dataset of 10,000 rgb images with corresponding binary masks (only one class). I have made the required changes to the preprocess function in dataset.py to ensure that the loaded masks are binary. I'm able to run train.py successfully but the results are not very encouraging.
> 
> 1. On training with a small subset of 50 images (40 training and 10 validation) with a learning rate of pow(10,-6), batch size of 2, scale factor of 1, up convolution of bilinear upscaling for 50 epochs. I see that the validation coefficient stays very low throughout (~order of pow(10,-3)). Batch loss is also very high (15 to 25). I'm not able to overfit the network on a small dataset.
> 2. On training with all 10,000 images with validation fraction = 20%, learning rate = pow(10,-5), epochs = 50, scale = 1 and batch size = 1,  validation dice coefficient reaches a maximum of 0.62 and keeps oscillating around 0.62. It doesn't go beyond 0.64. Batch loss fluctuates a lot between 0.2 and 0.7. The predicted results also don't look good.
> 
> Could you please suggest training hyperparameters like batch size, learning rate, epochs to be used for a dataset as big as 10,000 images to get good results? Also, what could be the reason for not being able to overfit on a small dataset?

Hi there, I'm curious if you were able to work this out and if you might be open to sharing how you did.  Thanks

**Comment by milesial** [comment]
Closing, as this issue is now outdated after the recent code refactor.
